core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -242,22 +243,39 @@ private boolean hasContentTypeOrCanAutoDetect(final RestRequest restRequest, fin
                 // be removed!
                 deprecationLogger.deprecated("Plain text request bodies are deprecated. Use request parameters or body " +
                     "in a supported format.");
-            } else if (isContentTypeRequired) {
-                return false;
-            } else {
-                deprecationLogger.deprecated("Content type detection for rest requests is deprecated. Specify the content type using " +
-                    "the [Content-Type] header.");
-                XContentType xContentType = XContentFactory.xContentType(restRequest.content());
-                if (xContentType == null) {
+            } else if (restHandler != null && restHandler.supportsContentStream() && restRequest.header("Content-Type") != null) {
+                final String lowercaseMediaType = restRequest.header("Content-Type").toLowerCase(Locale.ROOT);
+                // we also support line-delimited JSON, which isn't official and has a few variations
+                // http://specs.okfnlabs.org/ndjson/
+                // https://github.com/ndjson/ndjson-spec/blob/48ea03cea6796b614cfbff4d4eb921f0b1d35c26/specification.md
+                if (lowercaseMediaType.equals("application/x-ldjson") || lowercaseMediaType.equals("application/x-ndjson")) {
```

Comment:
Carriage return and newline also works:

```java
public void testSimpleBulkWithCarriageReturn() throws Exception {
        String bulkAction = "{ \"index\":{\"_index\":\"test\",\"_type\":\"type1\",\"_id\":\"1\"} }\r\n{ \"field1\" : \"value1\" }\r\n";
        BulkRequest bulkRequest = new BulkRequest();
        bulkRequest.add(bulkAction.getBytes(StandardCharsets.UTF_8), 0, bulkAction.length(), null, null, XContentType.JSON);
        assertThat(bulkRequest.numberOfActions(), equalTo(1));
        assertThat(((IndexRequest) bulkRequest.requests().get(0)).source(), equalTo(new BytesArray("{ \"field1\" : \"value1\" }\r")));
        Map<String, Object> sourceMap = XContentHelper.convertToMap(((IndexRequest) bulkRequest.requests().get(0)).source(),
            false, XContentType.JSON).v2();
        assertEquals("value1", sourceMap.get("field1"));
    }
```

The support for this could be cleaned up separately to trim the carriage return from the source.